TAS: KEP update: Add Relaxed fit algorithm to TAS (#4542)
kubernetes-sigs/kueue: keps/2724-topology-aware-scheduling/README.md

  // at the highest topology level, then it gets admitted as distributed
  // among multiple topology domains.
  PodSetPreferredTopologyAnnotation = "kueue.x-k8s.io/podset-preferred-topology"

>   // PodSetRelaxedTopologyAnnotation indicates that a PodSet requires

Reviewer: Maybe something more direct like "No topology required"?

Author: Done

>   // Topology Aware Scheduling, but it promotes filling up nodes to maximum over

Reviewer: It is mixing up pod needs with cluster efficiency. Can we have these things separated. Annotation on a podset should mean that Kueue is free to place podset as it see fit in a given moment, and the current implementation is to fill the already utilozed nodes (but may change in the future).

Author: Done

  // compact placement. The pods have lower chance of being scheduled on the same
  // node, but it mitigates resource fragmentation, and can lead to better

>   // node utilization. Recommended for PodSets that don't require heave inter pod

Reviewer: Fix wording around "don't require heave".

Author: Done

  // at the highest topology level, then it gets admitted as distributed
  // among multiple topology domains.
  PodSetPreferredTopologyAnnotation = "kueue.x-k8s.io/podset-preferred-topology"

>   // PodSetUnconstrainedTopologyAnnotation indicates that a PodSet requires does not have any topology requirements.

Reviewer 2: ```suggestion
  // PodSetUnconstrainedTopologyAnnotation indicates that a PodSet does not have any topology requirements.
```

Author: Done

>   // Kueue admits the PodSet if there's enough free capacity anywhere in the cluster.

Reviewer 2: ```suggestion
  // Kueue admits the PodSet if there's enough free capacity available.
```
not necessarily in the whole cluster. This is still constrained by the RF.spec.nodeLabels.

  // Recommended for PodSets that don't require pod-to-pod communication

>   PodSetUnconstrainedTopologyAnnotation = "kueue.x-k8s.io/podset-unconstrained-topology"

Reviewer 3: What are the differences between this annotation and one without any annotations?

Reviewer 3: Additionally, what is expected value? Bool?

Author: Yes, bool is expected value. The difference between this annotation and no annotation is that with TAS we mitigate the risk of admitting a job when capacity is fragmented. Imagine you have 2 GPUs scattered across 2 nodes, and you want to schedule a single pod that requires. Without TAS (and this annotation) a job would be falsely admitted even it cannot run.

A natural question may appear: why not setting just `preferred:hostname`? The difference between those two is that with preferred TAS try to pack pods as tighly as possible. While it's valuable for workloads that require heave pod-to-pod communication, that's not always the case, and a user might prefer to place pods on nodes scattered across the whole datacenter e.g. to reduce fragmentation

Author: Added clarification

Reviewer 3: Thank you. I added a clarified question to another thread.

  // at the highest topology level, then it gets admitted as distributed
  // among multiple topology domains.
  PodSetPreferredTopologyAnnotation = "kueue.x-k8s.io/podset-preferred-topology"

  // PodSetUnconstrainedTopologyAnnotation indicates that a PodSet requires no topology.
  // Kueue admits the PodSet if there's enough free capacity anywhere in the cluster.

>   // Recommended for PodSets that don't require inter pod communication

Reviewer 2: ```suggestion
  // Recommended for PodSets that don't require pod-to-pod communication
```
maybe makes it more clear

Author: Done

  PodSetUnconstrainedTopologyAnnotation = "kueue.x-k8s.io/podset-unconstrained-topology"
)
```

> The annotations are mutually exclusive and shouldn't be combined.

Reviewer 2: I would suggest to move this note to the Validation section, as we already validate against it.

Author: Done

  // at the highest topology level, then it gets admitted as distributed
  // among multiple topology domains.
  PodSetPreferredTopologyAnnotation = "kueue.x-k8s.io/podset-preferred-topology"

  // PodSetUnconstrainedTopologyAnnotation indicates that a PodSet does not have any topology requirements.
  // Kueue admits the PodSet if there's enough free capacity available.
  // Recommended for PodSets that don't need low-latency or high-throughput pod-to-pod communication,
  // but want to leverage TAS capabilities improve accuracy of admitting jobs
  //
  // +kubebuilder:validation:Type=boolean

>   PodSetUnconstrainedTopologyAnnotation = "kueue.x-k8s.io/podset-unconstrained-topology"

Reviewer 3: Does this mean that all Pods need to have either annotations (`podset-required-topology`, `podset-preferred-topology` or `PodSetUnconstrainedTopologyAnnotation`)?

My previous question is what are differences no TAS annotation and `PodSetUnconstrainedTopologyAnnotation `.
It looks like both (no TAS annotation and PodSetUnconstrainedTopologyAnnotation) are same.

Author: In Kueue v0.10.0 PodSet needs to have one of annotation (required/preferred) in order to use TAS. The same goes with the unconstrained annotation, if a PodSet wants to leverage TAS it needs to have some annotation defines. So the difference between no annotation and the unconstrained annotation would be to whether or not use TAS. Using TAS even if there are no heavy pod-to-pod communication might be beneficial to mitigate resource fragmentation.

Please note that in Kueue v0.11.0 mimowo introduces a mechanism to implicitly default the annotation (to `preferred`: [lowest level]) if all RFs in a CQ has Topology defined. This KEP takes it into account, as we allow a user to change the implicit default (to `unconstrained`)

Author: /hold cancel
Thank you!

- if "podset-required-topology" or "podset-preferred-topology" is specified for
  one PodTemplate, then one of them is specified for every PodTemplate in the
  Job spec.
- the annotations `kueue.x-k8s.io/podset-required-topology`,
  `kueue.x-k8s.io/podset-preferred-topology`, and `kueue.x-k8s.io/podset-unconstrained-topology`

>   are mutually exclusive and shouldn't be combined.

Author: Done

  // +optional
  Preferred *string `json:"preferred,omitempty"`

>   // Unconstrained indicates the topology assignment for the PodSet should be,

Reviewer 2: ```suggestion
  // Unconstrained indicates the topology assignment for the PodSet should be,
```
Should be sounds like it is not necessarily. Can we rephrase as  "is" or "is by default"?

Author: Done

>   // computed using the Unconstrained algorithm. Indicated by the

Reviewer 2: IIUC we don't have "Unconstrained algorithm", but in any case, I think we can assume the algorithm is not tightly coupled to the annotation definition.

Author: Done

  // +optional
  Preferred *string `json:"preferred,omitempty"`

>   // Unconstrained indicates that Kueue attempts to fit the PodSet within the entire available capacity,
>   // disregarding whether the nodes belong to the same domain.

Reviewer 2: ```suggestion
  // unconstrained indicates that Kueue has freedom to schedule the PodSet within
  // the entire available capacity, regardless of domain placement.
```
nit, proposal to rephrase.

Author: Done

topology level defined in the Topology referenced by the selected TAS flavor.
We call it "implicit default" as the annotation isn't persisted.

To change the implicit default from `podset-preferred-topology` to `podset-unconstrained-topology`

> a user can switch on the `TASImplicitDefaultUnconstrained` feature gate.

Reviewer 2: Please add somewhere a note that the feature gate is "deprecated" and only for experimentation to collect user feedback. If users find this configuration valuable we will introduce a dedicated TAS configuration.

Author: Done

  level. Kueue starts the search from the specified level, but if the PodSet
  does not fit, then it tries higher levels in the hierarchy.

Kueue packs pods on domains starting from the domains with the most free capacity. However, Kueue can operate in three modes when it comes to choosing the last domain if there is more than one capable of accommodating the remaining pods:
- `MostAllocated` - Kueue chooses the domain with the least available resource that is capable of accommodating all the pods to mitigate resource fragmentation
- `LeastAllocated` - Kueue chooses the domain that has the most available resources, providing better nodes utilization

> - `ConditionalBestFit` - Kueue chooses the domain with the least available resources. It promotes minimizing resource fragmentation, and higher node utilization over compact placement. This algorithm is used for PodSet with the `kueue.x-k8s.io/podset-relaxed-topology: true` annotation set.

Reviewer 2: tbh I dont get neither the Conditional not BestFit parts of the name :) . Maybe it should be called 
LeastFreeCapacityFirst?

and to be fair the current names are not good either. I would call them 'MostFreeCapacityFirst' and 'BestFit'.

Reviewer 2: > Are you proposing:
> MostAllocated -> BestFit
> LeastAllocated -> MostFreeCapacityFirst
> ConditionalBestFit -> LeastFreeCapacityFirst

That's right.

Author: Done

By default Kueue uses the `MostAllocated` algorithm. To use `LeastAllocated` algorithm, a user needs to set the feature gate `TASLeastAllocated` to `true`.

> To use `ConditionalBestFit` algorithm, a PodSet needs to have the `kueue.x-k8s.io/podset-relaxed-topology: true` annotation set. To default this annotation for all Workloads in the cluster a user can set the feature gate `TASImplicitDefaultRelaxed` to `true`.

Reviewer 2: Regarding the feature gate, it does not really default the annotation. It just makes kueue assume it as an implicit default when explicit annotation is not specified.

  level. Kueue starts the search from the specified level, but if the PodSet
  does not fit, then it tries higher levels in the hierarchy.

Kueue packs pods on domains starting from the domains with the most free capacity. However, Kueue can operate in three modes when it comes to choosing the last domain if there is more than one capable of accommodating the remaining pods:
- `MostAllocated` - Kueue chooses the domain with the least available resource that is capable of accommodating all the pods to mitigate resource fragmentation
- `LeastAllocated` - Kueue chooses the domain that has the most available resources, providing better nodes utilization

> - `Relaxed` - Kueue chooses the domain with least available resources. It promotes minimizing resource fragmentation, and higher node utilization over compact placement. This is signalized by the `kueue.x-k8s.io/podset-relaxed-topology: true` PodSet annotation.

Author: Done

  level. Kueue starts the search from the specified level, but if the PodSet
  does not fit, then it tries higher levels in the hierarchy.

> Kueue packs pods on domains with different algorithms, depending on the annotation and chosen profile:

Reviewer 2: ```suggestion
Kueue places pods on domains with different algorithms, depending on the annotation and chosen profile:
```
Since we have `LeastFreeCapacity` maybe "packs" is misleading.

Author: Done

- `MostFreeCapacity` algorithm - Kueue selects as many domains as needed (if it meets user's requirement) starting from the one with the most free capacity;
- `LeastFreeCapacity` algorithm - Kueue selects as many domains as needed (if it meets user's requirement) starting from the one with the least free capacity;
- `BestFit` algorithm - Kueue selects as many domains as needed (if it meets user's requirement) starting from the one with the most free capacity, but tries to optimize the last domain to leave as few free resources as possible.

Selection of the algorithm depends on TAS profiles expressed by feature gates, and PodSet's annotation:

| featuregate/annotation                  | preferred         | required          | unconstrained     |
| --------------------------------------- | ----------------- | ----------------- | ----------------- |
| None                                    | BestFit           | BestFit           | BestFit           |
| TASProfileMostFreeCapacity (deprecated) | MostFreeCapacity  | MostFreeCapacity  | MostFreeCapacity  |
| TASProfileMixed                         | BestFit           | BestFit           | LeastFreeCapacity |

> | TASProfileLeastFreeCapacity             | LeastFreeCapacity | LeastFreeCapacity | LeastFreeCapacity |

Reviewer 2: I would suggest to make these two feature gates deprecated too. The purpose of introducing them varies:
- `TASProfileMostFreeCapacity` - introduced as a bailout.
- `TASProfileMixed` - introduced for experimentation and collecting feedback
-  `TASProfileLeastFreeCapacity` - introduced for experimentation and collecting feedback

But I think all of them are "deprecated", because none of them is graduable in the current form.
I would suggest to:
- open an issue to introduce TAS configuration
- add a graduation criteria for TAS for beta to introduce the configuration

Author: Done

  level. Kueue starts the search from the specified level, but if the PodSet
  does not fit, then it tries higher levels in the hierarchy.

Kueue places pods on domains with different algorithms, depending on the annotation and chosen profile:
- `MostFreeCapacity` algorithm - Kueue selects as many domains as needed (if it meets user's requirement) starting from the one with the most free capacity;
- `LeastFreeCapacity` algorithm - Kueue selects as many domains as needed (if it meets user's requirement) starting from the one with the least free capacity;

> - `BestFit` algorithm - Kueue selects as many domains as needed (if it meets user's requirement) starting from the one with the most free capacity, but tries to optimize the last domain to leave as few free resources as possible.

Reviewer 3: If the topology is flat (only nodeLabel), what happens?

Reviewer 2: We don't know yet, but only BestFit sounds like a reasonable default.

We may still keep other algorithms, depending on user feedback, under some dedicated configuration (not feature gates), and the user feedback will help us to design the configuration. 

I think we could already open an issue for revisiting the TAS configuration regarding the placement algorithms.

Reviewer 3: Sorry what I wanted to ask is about "but tries to optimize the last domain to leave as few free resources as possible."

Does this "last domain" mean lowest level, right? In that case, this BesiFit algorithm behavior is mostly LeastFreeCapacity in the case of flat topology (only nodeLabel)?

Author: No, the last domain means the last domain to choose at any given level. Imagine you want to schedule 7 pods and nodes with free capacity that can fit 3,3,2 and 1 pod. In that case both `BestFit` algorithm and `MostFreeCapacity` algorithm choose first two nodes with 3 available pods. The difference occurs when selecting the last domain:
- `BestFit` will choose a node that can fit 1 pod, as it's enough
- `MostFreeCapacity` will choose a node that can fit 2 pods, as it's the one with most free capacity

Author: Clarified, and added example as well

Reviewer 3: Thank you, looks great.

Selection of the algorithm depends on TAS profiles expressed by feature gates, and PodSet's annotation:

| featuregate/annotation                   | preferred         | required          | unconstrained     |
| ---------------------------------------- | ----------------- | ----------------- | ----------------- |
| None                                     | BestFit           | BestFit           | BestFit           |
| TASProfileMostFreeCapacity (deprecated)  | MostFreeCapacity  | MostFreeCapacity  | MostFreeCapacity  |
| TASProfileMixed (deprecated)             | BestFit           | BestFit           | LeastFreeCapacity |

> | TASProfileLeastFreeCapacity (deprecated) | LeastFreeCapacity | LeastFreeCapacity | LeastFreeCapacity |

Reviewer 3: So, eventually, do we want to supply only BestFit, right?

Reviewer 2: @PBundyra may want to add something, but my understanding is summarized in https://github.com/kubernetes-sigs/kueue/pull/4542#discussion_r1991039901. Basically BestFit is the only algorithm which works sensibly as a default. However, other algorithms may find some niche applicability, so we keep the feature gates for experimentation. We don't plan to graduate the feature gates, but remove. We either remove the feature gates along with the algorithms, or introduce a dedicated configuration for them (TBD).

Reviewer 3: Uhm, I see. That makes sense. I just want to leave a note in the KEP that we currently expect the BestFit works well in most situations, but we might make the possibility for other deprecated scheduling based on user feedback.

Author: +1 to what mimowo said. I'll add clarification

> Feature gates: `TASProfileLeastAllocated`, `TASProfileMixed` and `TASProfileLeastFreeCapacity` are mutually exclusive.

Reviewer 2: sgtm, I don't have a strong view, just wasn't sure what it means in this context.
